### PR TITLE
Add new module: materialized sql view

### PR DESCRIPTION
**Pull request by petrus-v**
Hi,

I'm Pierre Verkest from Anybox, I'm pretty new in OCA and not sure about
the workflow to add new modules. Anyway this is a module we developed
on [bitbucket](https://bitbucket.org/anybox/materialized_sql_view).

I'm glad to contribute and dive into OCA tools and thanks @legalsylvain who
ask us to push this `materialized_sql_view` module here.

If you want to know more about this module you can read materialized_sql_view
module description in `__openerp__.py`.

I had to separate tests in an other module as I would tested the module with a
real model but wouldn't see it on production servers.

I already prepare a [branch running under v7](https://github.com/petrus-v/server-tools/tree/7.0-materialized_sql_view)
I guess if it's like odoo workflow I haven't to make a PR on both branches (7.0 and 8.0) ?

I was surprised some tested pass in Travis, like this one:
 https://travis-ci.org/petrus-v/server-tools/builds/46459751 based on this
commit: https://github.com/petrus-v/server-tools/commit/ab614eddd7f11dc81251080960f1969e6454d396
which should failed.

regards,

### CODE REVIEW

File: materialized_sql_view/__openerp__.py
Comment (clonedagain) [inline materialized_sql_view/__openerp__.py]: I think 0.1 would suffice.

**Comment by clonedagain** [comment]
That looks nice. I wonder if you couldn't simplify the code by dropping the legacy postgresql option. PG 9.3 has been available for quite some time and is the version recommanded by odoo SA.

**Comment by petrus-v** [comment]
@clonedagain,
Yes it could be easier if we had to support only one PG version, but we made it because we need it for our customer. Unfortunately we don't manage hosting for him, and it can take a while to make a plan with three parts (customer/hoster/IT integrator).

I could say with experience, companies (mainly true bigger they are) are afraid about change when we talk about their tool whose all employees depends so I'm sure I will find PG9.2 on servers for a couple years!

As you can see, if the PG cluster version is upgraded the module take care about it and will drop and recreate the view against the good version.

I hope the effort to maintain multi PG version will be benefit for the community we have also ideas to enhance this module which could explain some choices: 
- PG9.4 which was released in December introduce a new option `CONCURRENTLY` which could be easily introduce in this module by patterns we had chosen. 
- [using different strategies to refresh materialized view](http://tech.jonathangardner.net/wiki/PostgreSQL/Materialized_Views)
- Display on the UI the last refresh date until data was collected
- Provide a dependency refresh engine (in case of some materialized view depend of some other)

Thanks for your review, I already push the version to 0.1 as recommended.

**Comment by petrus-v** [comment]
I had not seen that @legalsylvain made the issue #101 before asking us by email which is related to this PR.

File: materialized_sql_view/__openerp__.py
Comment (yvaucher) [inline materialized_sql_view/__openerp__.py]: In v8.0 description can be put in README.rst to be visible both in github and on Odoo

You can find a sample here:
https://github.com/OCA/maintainers-tools/blob/master/template/module/README.rst

**Comment by yvaucher** [comment]
fr.po is there but pot file is missing

**Comment by hbrunn** [comment]
@clonedagain current CentOS ships with postgresql 9.2, so I think it's definitely worthwhile to support that. The code looks good to me, after the README.rst issue and the missing pot file is fixed, I think this is good to merge.

**Comment by legalsylvain** [comment]
:+1: (Code review, No test.)

**Comment by petrus-v** [comment]
Sorry I'm quite busy actually, sure I will take in consideration your advices. I also have to push a fixe that we figure out using it on production (a use case when it's crash by space missing on the Postgresql temp directory partition ) !

I'll do it this week!

**Comment by petrus-v** [comment]
My apologies to reviewers I've just realized it's probably not fair to add capabilities (commit cbbdc16) after your reviews on the module.

File: materialized_sql_view/__openerp__.py
Comment (yvaucher) [inline materialized_sql_view/__openerp__.py]: Thanks for having created README.rst

Now you just need to remove `description` key from `__openerp__.py`

**Comment by clonedagain** [comment]
@petrus-v I think it's OK to improve before merging, as long as you notify the reviewers. I prefer that than a new PR.

**Comment by yvaucher** [comment]
Restarted 3rd job of travis. Due to an error when pulling dependencies:

```
ImportError: No module named pytz
```

Not the first time I see this.

**Comment by yvaucher** [comment]
BTW @petrus-v don't mind squashing those pep8 / pylint commit. That doesn't help much in history.

**Comment by petrus-v** [comment]
I've just rebase the PR to have a look what's happen!

**Comment by StefanRijnhart** [comment]
Looks good (no test). Is travis just not completing its job or has it just been restarted?

File: materialized_sql_view/__openerp__.py
Comment (nbessi) [inline materialized_sql_view/__openerp__.py]: Please put OCA in the author as second author

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: I guess we can provide a base implementation here that can be call with super in inherited class

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: Just a matter of taste but I would prefere a None as default value

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: if the property are mandatory we may use the property decorator to raise a explicit message to map safe_properties

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: It is a technical error you should raise a ValueError not an osv exception.
Also osv.except_osv is deprecated

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: Are we sure for the cascading. It seems dangerous

File: materialized_sql_view/model/materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/materialized_sql_view.py]: Please use new API for new addons

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py]: Please use new API for new addon

File: materialized_sql_view/model/abstract_materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/abstract_materialized_sql_view.py:319]: why this syntax is it faster than {} constructor or a matter of taste? Personnaly I prefer the other syntax but it will not be blocking

File: materialized_sql_view/model/materialized_sql_view.py
Comment (nbessi) [inline materialized_sql_view/model/materialized_sql_view.py]: I missed something here what is the use of delegate_user?

**Comment by petrus-v** [comment]
I created PR #174 as this module was formerly develop against version 7.0, I'm still working on moving to the new API!

**Comment by guewen** [comment]
Looks great!
@petrus-v  Would you agree if we add an option to choose between a view and a materialized view? (then should the module have a more generic name?)

**Comment by petrus-v** [comment]
Hi @guewen,

If I understand you want to let admin to choose if the sql view is materialized at run time? It could be useful! :+1: 

So how should rename the module ? `base_sql_view`, `advanced_sql_view`, ... ?

Still some work to do here... I Just push what I started at Odoo Experiences, thanks again to @nbessi reviews... I know it's still not working but I want to make sure to not lost this work. I plan to carry on next PyCon meeting in october.

regards

**Comment by guewen** [comment]
@petrus-v I worked on an independent implementation in the meantime: https://github.com/OCA/server-tools/pull/245

**Comment by petrus-v** [comment]
@guewen it looks useful for customers, I catch your interest to add that option!

**Comment by legalsylvain** [comment]
Hi @petrus-v, 
Last year I realized a custom module that implement materialized view for V7.0 serie. [see module](https://github.com/grap/odoo-addons-misc/blob/7.0/pos_sale_reporting/model/materialized_model.py)
I'm currently working on porting all my modules in V8.0. I'd like to use 'standard' / 'community' module for that feature.
How can I help you to finish this PR ?

kind regards.

**Comment by petrus-v** [comment]
Hi @legalsylvain,

Sorry about being so slow on that! Do not hesitate to get that branch and fix it! 

Otherwise If you aren't hurry, I though to carry on this project at OCA code sprint at [PynConFr](http://www.pycon.fr/2015/) in october. Are you comming ?

Can you let me know if you do it, or any one else who want to do it?

File: materialized_sql_view/model/materialized_sql_view.py
Comment (codingforfun) [inline materialized_sql_view/model/materialized_sql_view.py:43]: Typo: I think you mean 'async' here.

**Comment by gurneyalex** [comment]
@petrus-v @legalsylvain looks nice. Could you check the crash in the cron on runbot? 

```
Programming error: can't adapt type 'ir.model', in query  SELECT "ir_model"."state","ir_model"."create_date","ir_model"."avoid_quick_create","ir_model"."info","ir_model"."id","ir_model"."model","ir_model"."write_uid","ir_model"."name","ir_model"."create_uid","ir_model"."write_date" FROM "ir_model"
                        WHERE "ir_model".id IN %s  ORDER BY "ir_model"."model"  
```

**Comment by laetitia-gangloff** [comment]
Hello @petrus-v ,
I start to use this module in V8 after success with V7. And it seems materialized view is always recomputed :( And I am sure this time this is an unicode materialized view.

@gurneyalex I get this error too, so I replace the line 77 from file `materialized_sql_view/model/materialized_sql_view.py` by `model = matview.model_id.model`
I will propose PR to @petrus-v with this change.

**Comment by petrus-v** [comment]
@laetitia-gangloff , thanks for the PR!

As far as I know we are not using this module on V8... The work still in progress and I'm very slow, sorry about that!

If someone want to carry on this migration, feel free to do it!

**Comment by Cedric-Pigeon** [comment]
Hi, I tried to use the refresh button of the materialized view.

The view seemed to be refreshed but the state of the view didn't change and stayed on "Refreshing"

**Comment by Cedric-Pigeon** [comment]
@petrus-v Do you plan to finish this PR ? Thanks.

**Comment by petrus-v** [comment]
@Cedric-Pigeon I've done some works last code sprint!

I'm gonna push it right now, helps is welcome however! Today I only use version 7!

File: test_materialized_sql_view/tests/test_abstract_materialized_sql_view.py
Comment (gurneyalex) [inline test_materialized_sql_view/tests/test_abstract_materialized_sql_view.py:73]: use 
```
sql = self.demo_matview_mdl.sql_view_definition
```

File: test_materialized_sql_view/tests/test_abstract_materialized_sql_view.py
Comment (gurneyalex) [inline test_materialized_sql_view/tests/test_abstract_materialized_sql_view.py:13]: line too long (79 char max)

**Review by gurneyalex — CHANGES_REQUESTED** [review]
could you fix the two small issues and rebase the PR?

**Comment by pedrobaeza** [comment]
This has been stalled for a long time. It's an interesting idea, but it seems unfinished. If someone wants to revive this, please open a new PR.

**Comment by legalsylvain** [comment]
materialized view can now be generated by bi_sql_editor OCA module.

**Pull request closed without merging**
